ipaserver/install/ipa_server_certinstall.py

```diff
@@ -147,21 +149,35 @@ def install_dirsrv_cert(self):
                                ['nssslpersonalityssl'])
         old_cert = entry.single_value['nssslpersonalityssl']
 
-        server_cert = self.import_cert(dirname, self.options.pin,
-                                       old_cert, 'ldap/%s' % api.env.host,
-                                       'restart_dirsrv %s' % serverid)
+        nickname, cert = self.import_cert(dirname, self.options.pin,
+                                          old_cert, 'ldap/%s' % api.env.host,
+                                          'restart_dirsrv %s' % serverid)
 
-        entry['nssslpersonalityssl'] = [server_cert]
+        entry['nssslpersonalityssl'] = [nickname]
         try:
             conn.update_entry(entry)
         except errors.EmptyModlist:
             pass
 
+        ds = dsinstance.DsInstance()
+        ds.suffix = api.env.basedn
+        ds.realm = api.env.realm
+        ds.fqdn = api.env.host
+        ds.service_prefix = 'ldap'
+        ds.cert = cert
+        name = f'{ds.service_prefix}/{ds.fqdn}'
+        oldcerts = api.Command.cert_find(service=name)['result']
+        ds.add_cert_to_service(append=False)
+
+        revoke_certs(oldcerts)
```

Comment:
I am not comfortable with the revocation of the previous cert. For instance, if the same 3rd-party cert is used initially for both LDAP and HTTP but later on the administrator realizes he should reduce the attack surface and replaces the LDAP cert with another one (in order to have a distinct cert for each service), he doesn't want to revoke the HTTP cert.
I know that 3rd part certs can't be revoked inside IPA but that's just an example for a scenario where replacing a cert is not synonym of revoking a cert.

Reply:
But they should be able to issue a new HTTP, LDAP or PKINIT cert from IPA using certmonger.   I'll give that a try but I don't see why it wouldn't work.

Reply:
So yes, it's possible. I tested replacing the Apache and LDAP certs with 3rd party certs, then replaced those with IPA-issued certificates.

For Apache you first need to move the cert and key out of the way (or remove them). Then you can have certminger issue a new one:

ipa-getcert request -f /var/lib/ipa/certs/httpd.crt  -k /var/lib/ipa/private/httpd.key -p /var/lib/ipa/passwds/ipa.example.test-443-RSA -D ipa.example.test -D ipa-ca.example.test -K HTTP/ipa.example.test@EXAMPLE.TEST -C /usr/libexec/ipa/certmonger/restart_httpd -T caIPAserviceCert -v -w

Restart httpd: systemctl restart httpd

The paths are the same for all HTTP certs, CA-issued or 3rd party

For 389-ds you can request a new cert:

ipa-getcert request -d /etc/dirsrv/slapd-EXAMPLE-TEST -n Server-Cert -p /etc/dirsrv/slapd-EXAMPLE-TEST/pwdfile.txt  -D ipa.example.test -K ldap/ipa.example.test@EXAMPLE.TEST -C "/usr/libexec/ipa/certmonger/restart_dirsrv EXAMPLE-TEST" -T caIPAserviceCert -v -w

Stop dirsrv

Edit dse.ldif and replace nsSSLPersonalitySSL with Server-Cert, or whatever nickname was chosen, then restart dirsrv.

Done and the certs are replaced and tracked now.

Reply:
In fact, I blogged about this a few years ago https://rcritten.wordpress.com/2019/05/07/how-do-i-revert-back-to-using-ipa-issued-web-ldap-certs/